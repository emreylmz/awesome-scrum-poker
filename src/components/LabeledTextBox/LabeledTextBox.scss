@import "../../styles/colors";
@import "../../styles/metrics";

.text-box {
  display: flex;
  flex-direction: row;
  flex: 0;
  align-content: center;

  &__label {
    flex: 1;
    text-align: right;
    margin: $default-margin
  }

  input {
    flex: 2;
    margin: $default-margin;

    font-size: 18px;
    padding: 10px;
    background-color: aliceblue;
    color: $primary-color;
    border-radius: $default-border-radius;
  }
}
