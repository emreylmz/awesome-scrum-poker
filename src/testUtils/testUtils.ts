export const emptyTestFunc = (): void => void 0
