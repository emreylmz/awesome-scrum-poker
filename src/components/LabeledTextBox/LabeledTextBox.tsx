import React from 'react'
import { Render } from '../../types'
import './LabeledTextBox.scss'

interface Props {
  label: string
  value?: string
  onChange?: (value?: string) => void
  characterLimit?: number
  numeric?: boolean
  validator?: (value?: string | number) => boolean
}

interface State {
  value?: string
  isValid: boolean
}

class LabeledTextBox extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      value: props.value,
      isValid: true
    }
  }

  UNSAFE_componentWillReceiveProps(nextProps: Props): void {
    if (nextProps.value !== this.props.value) {
      this.setState({ value: nextProps.value })
    }
  }

  private onValueChange = (e: any): void => {
    const { validator } = this.props
    const { value } = e.target
    const isValid = validator ? validator(value) : true

    this.setState(
      {
        value,
        isValid
      },
      () => {
        if (this.props.onChange) {
          this.props.onChange(value)
        }
      }
    )
  }

  render(): Render {
    const { label, characterLimit, numeric } = this.props
    const { value, isValid } = this.state

    let errorMessage = null
    if (!isValid) {
      errorMessage = <div>Ooops!</div>
    }
    return (
      <div className={'text-box'}>
        <div className={'text-box__label'}>{label}</div>
        <input
          value={value}
          maxLength={characterLimit}
          type={numeric ? 'number' : 'text'}
          onChange={this.onValueChange}
        />
        {errorMessage}
      </div>
    )
  }
}

export type LabeledTextBoxProps = Props
export default LabeledTextBox
