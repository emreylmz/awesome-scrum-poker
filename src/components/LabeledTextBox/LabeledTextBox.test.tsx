import { mount, shallow, ShallowWrapper } from 'enzyme'
import React from 'react'
import '../../testUtils/setupTests'
import LabeledTextBox, { LabeledTextBoxProps } from './LabeledTextBox'

type DoneCallback = jest.DoneCallback

const defaultProps: LabeledTextBoxProps = {
  label: 'Labeled Text Box'
}

const getWrapper = (props?: LabeledTextBoxProps): ShallowWrapper => {
  return shallow(<LabeledTextBox {...defaultProps} {...props} />)
}

describe('Labeled Text Box UI', () => {
  it('should render correctly', (done: DoneCallback) => {
    const wrapper = getWrapper()
    expect(wrapper).toMatchSnapshot()
    done()
  })

  it('should change its value with props', (done: DoneCallback) => {
    const newValue = 'input value'
    const wrapper = getWrapper()
    let textBox = wrapper.find('input')

    expect(textBox.prop('value')).toBeUndefined()
    wrapper.setProps({
      value: newValue
    })
    textBox = wrapper.find('input')
    expect(textBox.prop('value')).toEqual(newValue)
    done()
  })

  it('should have a label', (done: DoneCallback) => {
    const wrapper = getWrapper()
    const labelWrapper = wrapper.find('.text-box__label')
    expect(labelWrapper).toBeDefined()
    expect(labelWrapper.text()).toEqual(defaultProps.label)
    done()
  })

  it('should change label title with props', (done: DoneCallback) => {
    const testLabel = 'Test Label'
    const wrapper = getWrapper()
    let labelWrapper = wrapper.find('.text-box__label')
    expect(labelWrapper.text()).toEqual(defaultProps.label)
    wrapper.setProps({ label: testLabel })
    labelWrapper = wrapper.find('.text-box__label')
    expect(labelWrapper.text()).toEqual(testLabel)
    done()
  })
})

describe('Labeled Text Box Logic', () => {
  it('should be able to limit character', () => {
    const characterLimit = 200
    const wrap = mount(<LabeledTextBox {...defaultProps} characterLimit={characterLimit} />)
    const textBox = wrap.find('input')
    const newValue = new Array(300).fill('a').join('')

    const limitedValue = newValue.slice(0, characterLimit)
    expect(textBox.prop('maxLength')).toEqual(characterLimit)
    // wrap.setProps({ value: newValue })
    //
    // textBox = wrap.find('input')
    // expect(textBox.prop('value')).toEqual(limitedValue)

    // TODO: find a solution for test max length
  })

  it('should call onChange method when its value change', () => {
    // const renderer = getDefaultRenderer()
    const newValue = 'NEW VALUE'
    const onChange = (value?: string | number): void => {
      expect(value).toEqual(newValue)
    }
    const wrap = mount(<LabeledTextBox {...defaultProps} onChange={onChange} />)
    // spyOn(wrap.prop())
  })
})
