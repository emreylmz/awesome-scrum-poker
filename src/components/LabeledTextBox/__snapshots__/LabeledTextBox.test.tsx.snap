// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Labeled Text Box UI should render correctly 1`] = `ShallowWrapper {}`;
